### PR TITLE
Create Universal Query Language

### PR DESCRIPTION
**Pull request by PawelPeczek-Roboflow**
# Description

In this PR I ship extension to core block in form of universal query / operation language.

The idea is the following:
* we define JSON-based language to express operations on different types of data that we support (for example, for `sv.Detections` we support filtering, offseting and getting detections metadata - like size or location) and evaluation of unary and binary operations that are useful for conditional logic
* definition is extensible - to add new operation we only need to create pydantic entity to host its definition, implementation of that operation and register the operation in UQL execution engine
* definitions are self-describing - entities defining ops expose JSON manifests including custom fields expressing types of operations - that together should allow creating relatively generic UI for different kinds of blocks using UQL and execute simple **type inference** procedure on UI side to very effectively narrow down allowed operations that users can apply in specific contexts - hopefully making the whole thing quite intuitive
* we do not make changes into `workflows` EE, apart from quality of life improvements - *selectors embedded in blocks manifests dicts* (so far we supported only singular references and lists of references, now we would like to be able to provide Dict[str, reference] as input - as this is very useful for providing parametrisation of UQL execution engine - let's say that u want to filter detections, but threshold of confidence is provided in runtime, not hardcoded in definition)
* I do not imply that all blocks should use UQL, this is optional breadth-wise extension for workflows, but I think it will be beneficial, as:
  * we would be able to have quite generic builders for tons of operations in UI
  * we would be able to quickly extend UQL
  * we would eliminate need for multiple custom glue code blocks and python-code blocks (to some extent)

Downsides of the idea:
* there needs to be some intelligence in the UI to create builders for operations and to know how to behave in context of specific blocks (for instance - condition would provide all parameters of UQL EE in the manifest definition vs detections filter would implicitly iterate over predictions in batch that is provided as `predictions` parameter) 
* the JSON definition is not quite easy to read by human - but it is quite easy to programatically construct and parse, which is more important in that case. I do recommend not pushing into "verbose" UQL, as we will add significant amount of work on the parsing side and on the query construction side, as both sides would need to implement "parser" of some kind

Here are the examples:
* [Block with detections sampling (to be used for AL)](https://gist.github.com/PawelPeczek-Roboflow/fbf2c63374e75679868ebaa00bee8023)
* [Block with new condition logic](https://gist.github.com/PawelPeczek-Roboflow/7c3e1ea37253c9b1bf5eccc49df64740)
* [Block with transformation (filtering)](https://gist.github.com/PawelPeczek-Roboflow/efa7a417285e741039a63dbacef57056)

As u see, they all operate on the very similar definitions and in general this extension bring a lot of expression power.

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

* NOT READY TO MERGE

## Any specific deployment considerations

* We need `sv.Detections` as main data format to proceed with this PR (requires merging https://github.com/roboflow/inference/pull/392)
* we need to push changes started here further - AL Data collector needs to have new sink and we need to make the operation set better and more robust 

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: inference/core/workflows/core_steps/models/foundation/lmm_classifier.py [hunk 1]
```diff
@@ -156,11 +164,20 @@ async def run_locally(
         for p, i in zip(predictions, images):
             p[PREDICTION_TYPE_KEY] = "classification"
             p[PARENT_ID_KEY] = i[PARENT_ID_KEY]
-        return predictions
+        return images.align_batch_results(
+            results=predictions,
+            null_element={
+                "raw_output": None,
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/lmm_classifier.py:170]: where are those keys coming from?

File: inference/core/workflows/core_steps/models/foundation/ocr.py [hunk 2]
```diff
@@ -135,19 +147,35 @@ async def run_remotely(
...
+        results = self._post_process_result(
+            predictions=predictions,
+            images=non_empty_images,
+        )
+        return images.align_batch_results(
+            results=results,
+            null_element={
+                PARENT_ID_KEY: None,
+                ROOT_PARENT_ID_KEY: None,
+                "result": None,
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/ocr.py:166]: magic string

File: inference/core/workflows/core_steps/models/foundation/yolo_world.py [hunk 3]
```diff
@@ -187,25 +197,25 @@ async def run_remotely(
...
         self,
-        image: List[dict],
+        images: List[WorkflowImageData],
         predictions: List[dict],
     ) -> List[Dict[str, Union[sv.Detections, Any]]]:
-        detections = convert_to_sv_detections(predictions)
-        for p, d in zip(predictions, detections):
-            p["predictions"] = d
-        predictions = attach_prediction_type_info(
+        predictions = convert_inference_detections_batch_to_sv_detections(predictions)
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/foundation/yolo_world.py:212]: within inference project do we have another structure which should be converted to sv.Detections?

File: inference/core/workflows/core_steps/models/roboflow/instance_segmentation.py [hunk 4]
```diff
@@ -200,10 +198,14 @@ async def run_locally(
         disable_active_learning: Optional[bool],
         active_learning_target_dataset: Optional[str],
     ) -> List[Dict[str, Union[sv.Detections, Any]]]:
+        non_empty_images = [i for i in images.iter_nonempty()]
+        non_empty_inference_images = [
+            i.to_inference_format(numpy_preferred=True) for i in non_empty_images
+        ]
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/roboflow/instance_segmentation.py:204]: this pattern seems to be repeating across blocks, maybe `images.iter_nonempty` can take `numpy_preferred` as parameter and return required inference images in one go?

File: inference/core/workflows/core_steps/models/third_party/barcode_detection.py [hunk 5]
```diff
@@ -72,59 +72,52 @@ def get_manifest(cls) -> Type[WorkflowBlockManifest]:
...
+    class_name = np.array(class_name) if len(class_name) > 0 else np.empty(0)
+    detections = sv.Detections(
+        xyxy=np.array(xyxy),
+        confidence=confidence,
+        class_id=class_id,
+        data={CLASS_NAME_DATA_FIELD: class_name},
+    )
+    detections[DETECTION_ID_KEY] = np.array([uuid4() for _ in range(len(detections))])
+    detections[PREDICTION_TYPE_KEY] = np.array(
+        ["barcode-detection" for _ in range(len(detections))]
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/third_party/barcode_detection.py:115]: We could avoid loop: `detections[PREDICTION_TYPE_KEY] = np.array(["barcode-detection"]*len(detections))`

File: inference/core/workflows/core_steps/models/third_party/barcode_detection.py [hunk 6]
```diff
@@ -72,59 +72,52 @@ def get_manifest(cls) -> Type[WorkflowBlockManifest]:
...
+    )
+    detections["data"] = (
+        np.array(extracted_data) if len(extracted_data) > 0 else np.empty(0)
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/third_party/barcode_detection.py:118]: `extracted_data` is defined as list, so we can simplify to `detections["data"] = np.array(extracted_data) if extracted_data else np.empty(0)`

Also:

```
>>> np.array([])
array([], dtype=float64)
>>> np.empty(0)
array([], dtype=float64)
```

So we could further simplify to `detections["data"] = np.array(extracted_data)`

File: inference/core/workflows/core_steps/models/third_party/qr_code_detection.py [hunk 7]
```diff
@@ -71,62 +68,57 @@ def get_manifest(cls) -> Type[WorkflowBlockManifest]:
...
+        class_id=class_id,
+        data={CLASS_NAME_DATA_FIELD: class_name},
+    )
+    detections[DETECTION_ID_KEY] = np.array([uuid4() for _ in range(len(detections))])
+    detections[PREDICTION_TYPE_KEY] = np.array(
+        ["qrcode-detection" for _ in range(len(detections))]
+    )
+    detections["data"] = (
+        np.array(extracted_data) if len(extracted_data) > 0 else np.empty(0)
+    )
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/third_party/qr_code_detection.py:120]: Can be simplified to `detections["data"] = np.array(extracted_data)`

File: inference/core/workflows/core_steps/models/third_party/qr_code_detection.py [hunk 8]
```diff
@@ -71,62 +68,57 @@ def get_manifest(cls) -> Type[WorkflowBlockManifest]:
...
+    detections = sv.Detections(
+        xyxy=np.array(xyxy),
+        confidence=confidence,
+        class_id=class_id,
+        data={CLASS_NAME_DATA_FIELD: class_name},
+    )
+    detections[DETECTION_ID_KEY] = np.array([uuid4() for _ in range(len(detections))])
+    detections[PREDICTION_TYPE_KEY] = np.array(
+        ["qrcode-detection" for _ in range(len(detections))]
+    )
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/models/third_party/qr_code_detection.py:117]: Loop can be avoided: `detections[PREDICTION_TYPE_KEY] = np.array(["qrcode-detection"] * len(detections))`

File: inference/core/workflows/core_steps/transformations/detections_transformation.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    WorkflowImageSelector,
+    WorkflowParameterSelector,
+)
+from inference.core.workflows.prototypes.block import (
+    WorkflowBlock,
+    WorkflowBlockManifest,
+)
+
+LONG_DESCRIPTION = """
+"""
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/transformations/detections_transformation.py:35]: empty

File: inference/core/workflows/execution_engine/executor/output_constructor.py [hunk 10]
```diff
@@ -15,6 +19,9 @@ def construct_workflow_output(
     workflow_outputs: List[JsonField],
     execution_cache: ExecutionCache,
 ) -> Dict[str, List[Any]]:
+    # TODO: figure out if we needed generic "to-root coordinates" transformation?
+    # Maybe output constructor should never touch this? Maybe it's better to
+    # have steps transforming to specific coordinates systems?
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/execution_engine/executor/output_constructor.py:24]: should we extract this to separate task? It seems we are already handling this by calling `sv_detections_to_root_coordinates` below

File: inference/core/interfaces/http/orjson_utils.py [hunk 11]
```diff
@@ -117,70 +102,8 @@ def contains_image(element: Any) -> bool:
...
-                        KEYPOINTS_CLASS_ID_KEY: int(keypoint_class_id),
-                        KEYPOINTS_CLASS_NAME_KEY: str(keypoint_class_name),
-                        KEYPOINTS_CONFIDENCE_KEY: float(keypoint_confidence),
-                        X_KEY: float(x),
-                        Y_KEY: float(y),
-                    }
-                )
-        serialized_detections.append(detection_dict)
-    return serialized_detections
+def serialise_image(image: WorkflowImageData) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/http/orjson_utils.py:105]: 👍

File: inference/core/utils/image_utils.py [hunk 12]
```diff
@@ -419,6 +419,7 @@ def convert_gray_image_to_bgr(image: np.ndarray) -> np.ndarray:
 
 def np_image_to_base64(image: np.ndarray) -> bytes:
     """
+    TODO: This function is probably broken!
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/utils/image_utils.py:422]: 👍

File: inference/core/workflows/core_steps/common/query_language/entities/operations.py [hunk 13]
```diff
@@ -0,0 +1,760 @@
...
+                    BOOLEAN_KIND,
+                ],
+            ],
+            "output_kind": [BOOLEAN_KIND],
+        },
+    )
+    type: Literal["!="]
+
+
+class NumerGreater(BinaryOperator):
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/entities/operations.py:484]: 👍

File: inference/core/workflows/core_steps/common/query_language/entities/operations.py [hunk 14]
```diff
@@ -0,0 +1,760 @@
...
+                [INTEGER_KIND, FLOAT_KIND, FLOAT_ZERO_TO_ONE_KIND],
+                [INTEGER_KIND, FLOAT_KIND, FLOAT_ZERO_TO_ONE_KIND],
+            ],
+            "output_kind": [BOOLEAN_KIND],
+        },
+    )
+    type: Literal["(Number) >"]
+
+
+class NumerGreaterEqual(BinaryOperator):
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/entities/operations.py:499]: 👍

File: inference/core/workflows/core_steps/common/query_language/entities/operations.py [hunk 15]
```diff
@@ -0,0 +1,760 @@
...
+                [INTEGER_KIND, FLOAT_KIND, FLOAT_ZERO_TO_ONE_KIND],
+                [INTEGER_KIND, FLOAT_KIND, FLOAT_ZERO_TO_ONE_KIND],
+            ],
+            "output_kind": [BOOLEAN_KIND],
+        },
+    )
+    type: Literal["(Number) >="]
+
+
+class NumerLower(BinaryOperator):
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/entities/operations.py:514]: 👍

File: inference/core/workflows/core_steps/common/query_language/entities/operations.py [hunk 16]
```diff
@@ -0,0 +1,760 @@
...
+                [INTEGER_KIND, FLOAT_KIND, FLOAT_ZERO_TO_ONE_KIND],
+                [INTEGER_KIND, FLOAT_KIND, FLOAT_ZERO_TO_ONE_KIND],
+            ],
+            "output_kind": [BOOLEAN_KIND],
+        },
+    )
+    type: Literal["(Number) <"]
+
+
+class NumerLowerEqual(BinaryOperator):
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/entities/operations.py:529]: 👍

File: inference/core/workflows/core_steps/common/query_language/entities/operations.py [hunk 17]
```diff
@@ -0,0 +1,760 @@
...
+            "output_kind": [BOOLEAN_KIND],
+        },
+    )
+    type: Literal["(Sequence) is not empty"]
+
+
+class StaticOperand(BaseModel):
+    type: Literal["StaticOperand"]
+    value: Any
+    operations: List["AllOperationsType"] = Field(default_factory=lambda: [])
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/entities/operations.py:721]: 👍

File: inference/core/workflows/core_steps/common/query_language/entities/operations.py [hunk 18]
```diff
@@ -0,0 +1,760 @@
...
+
+
+class DynamicOperand(BaseModel):
+    type: Literal["DynamicOperand"]
+    operations: List["AllOperationsType"] = Field(default_factory=lambda: [])
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/entities/operations.py:726]: 👍

File: inference/core/workflows/core_steps/common/query_language/operations/detection/base.py [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+def extract_detection_property(
+    value: Any,
+    property_name: DetectionsProperty,
+    unwrap_method: SequenceUnwrapMethod,
+    execution_context: str,
+    **kwargs,
+) -> Any:
+    if isinstance(value, sv.Detections):
+        if unwrap_method is SequenceUnwrapMethod.FIRST:
+            value = next(value.__iter__())
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/operations/detection/base.py:47]: 👍

File: inference/core/workflows/core_steps/common/query_language/operations/detections/base.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+        raise InvalidInputTypeError(
+            public_message=f"Executing extract_detections_property(...) in context {execution_context}, "
+            f"expected sv.Detections object as value, got {value_as_str} of type {type(value)}",
+            context=f"step_execution | roboflow_query_language_evaluation | {execution_context}",
+        )
+    return PROPERTIES_EXTRACTORS[property_name](value)
+
+
+def filter_detections(
+    value: Any,
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/operations/detections/base.py:50]: 👍

File: inference/core/workflows/core_steps/common/query_language/operations/detections/base.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+    DetectionsProperty.Y_MIN: lambda detections: detections.xyxy[:, 1].tolist(),
+    DetectionsProperty.X_MAX: lambda detections: detections.xyxy[:, 2].tolist(),
+    DetectionsProperty.Y_MAX: lambda detections: detections.xyxy[:, 3].tolist(),
+    DetectionsProperty.CLASS_ID: lambda detections: detections.class_id.tolist(),
+    DetectionsProperty.SIZE: lambda detections: detections.box_area.tolist(),
+}
+
+
+def extract_detections_property(
+    value: Any,
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/operations/detections/base.py:34]: 👍

File: inference/core/workflows/core_steps/common/query_language/operations/detections/base.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+            f"got {value_as_str} of type {type(value)}",
+            context="step_execution | roboflow_query_language_evaluation",
+        )
+    width_change = offset_x / 2
+    height_change = offset_y / 2
+    detections_copy = deepcopy(value)
+    detections_copy.xyxy[:, 0] -= width_change
+    detections_copy.xyxy[:, 2] += width_change
+    detections_copy.xyxy[:, 1] -= height_change
+    detections_copy.xyxy[:, 3] += height_change
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/operations/detections/base.py:86]: 👍

File: inference/core/workflows/core_steps/common/query_language/operations/detections/base.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+        raise InvalidInputTypeError(
+            public_message=f"Executing shift_detections(...), expected sv.Detections object as value, "
+            f"got {value_as_str} of type {type(value)}",
+            context="step_execution | roboflow_query_language_evaluation",
+        )
+    detections_copy = deepcopy(value)
+    detections_copy.xyxy[:, 0] += shift_x
+    detections_copy.xyxy[:, 2] += shift_x
+    detections_copy.xyxy[:, 1] += shift_y
+    detections_copy.xyxy[:, 3] += shift_y
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/operations/detections/base.py:102]: 👍

File: inference/core/workflows/core_steps/common/query_language/operations/core.py [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+    "ExtractDetectionProperty": extract_detection_property,
+    "DetectionsOffset": offset_detections,
+    "DetectionsShift": shift_detections,
+    "RandomNumber": generate_random_number,
+    "StringMatches": string_matches,
+    "SequenceLength": get_sequence_length,
+    "ExtractImageProperty": extract_image_property,
+    "Multiply": multiply,
+    "Divide": divide,
+}
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/operations/core.py:182]: no, that should be generated by UI builder

File: inference/core/workflows/core_steps/common/query_language/operations/utils.py [hunk 25]
```diff
@@ -0,0 +1,11 @@
...
+
+
+def safe_stringify(value: Any, max_characters: int = 128) -> str:
+    try:
+        str_value = str(value)
+        if len(str_value) > max_characters:
+            str_value = f"{str_value[:max_characters]} [...]"
+        return str_value
+    except Exception:
+        return "could not get string representation of value"
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/operations/utils.py:11]: yes

File: inference/core/workflows/core_steps/common/serializers.py [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+                        "y": float(y),
+                    }
+                )
+        serialized_detections.append(detection_dict)
+    image_metadata = {
+        "width": None,
+        "height": None,
+    }  # TODO: this breaks the contract of
+    # standard inference, but to fix that problem, we would need sv.Detections to provide
+    # detection-level metadata.
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/serializers.py:92]: let's discuss how to solve that

File: inference/core/workflows/core_steps/models/foundation/yolo_world.py (hunk 3 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/yolo_world.py:212]: not at this moment

File: inference/core/workflows/core_steps/models/roboflow/instance_segmentation.py (hunk 4 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/roboflow/instance_segmentation.py:204]: 👎

File: inference/core/workflows/core_steps/models/third_party/barcode_detection.py (hunk 5 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/third_party/barcode_detection.py:115]: 👍

File: inference/core/workflows/core_steps/models/third_party/barcode_detection.py (hunk 6 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/third_party/barcode_detection.py:118]: 👍

File: inference/core/workflows/core_steps/models/third_party/qr_code_detection.py (hunk 7 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/third_party/qr_code_detection.py:120]: 👍

File: inference/core/workflows/core_steps/models/third_party/qr_code_detection.py (hunk 8 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/third_party/qr_code_detection.py:117]: 👍

File: inference/core/workflows/core_steps/transformations/detections_transformation.py (hunk 9 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/transformations/detections_transformation.py:38]: 👍

File: inference/core/workflows/execution_engine/executor/output_constructor.py (hunk 10 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/execution_engine/executor/output_constructor.py:24]: issue submitted

**Pull request merged**